File: ci/read_version.py
import tomllib as toml
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
CHANGELOG_FILE = ROOT / "CHANGELOG.md"
METADATA_FILE = ROOT / "metadata.toml"


class MetadataError(Exception):
    """Raised when the metadata is invalid."""

    pass


def read_version(metadata_file: Path) -> str:
    """Reads the project version from metadata_file."""

    if not metadata_file.exists():
        raise MetadataError(f"{metadata_file.name} not found")

    data = toml.loads(metadata_file.read_text())

    try:
        mod = data["mod"]
    except KeyError:
        raise MetadataError(f"No `mod` table found in {metadata_file.name}")

    try:
        version = mod["version"]
    except KeyError:
        raise MetadataError(f"No `mod.version` entry found in {metadata_file.name}")

    if not version or not isinstance(version, str):
        raise MetadataError(f"Invalid `mod.version` found in {metadata_file.name}")

    return version

File: ci/src/read_version.ts
import TOML from "smol-toml";
import { existsSync, readFileSync } from "node:fs";
import { METADATA_FILE } from "./project_files.ts";
import { MetadataSchema } from "./metadata_model.ts";

export class MetadataError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MetadataError";
  }
}

export function readVersion(metadataFile: string = METADATA_FILE): string {
  if (!existsSync(metadataFile)) {
    throw new MetadataError(`${metadataFile} not found`);
  }

  const toml = TOML.parse(readFileSync(metadataFile, "utf-8"));
  const { success, data, error } = MetadataSchema.safeParse(toml);

  if (!success) {
    throw new MetadataError(
      `Invalid or missing \`mod.version\` in ${metadataFile}: ${error}`,
    );
  }

  return data.mod.version;
}
